File: 0257 - Binary Tree Paths/Binary Tree Paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        result = []
        self.dfs(root, result, "")
        return result
        
        
    def dfs(self, root: TreeNode, result: List[str], path: str):
        if root.left is None and root.right is None:
            path += str(root.val)
            result.append(path)
            return
        
        if root is not None:
            if root.left is not None:
                self.dfs(root.left, result, path + str(root.val) + "->")
                
            if root.right is not None:
                self.dfs(root.right, result, path + str(root.val) + "->")

File: 0257 - Binary Tree Paths/Binary Tree Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> result;
        dfs(root, "", result);
        return result;
    }
    
    void dfs(TreeNode* root, string path, vector<string>& result) {
        if (root->left == NULL && root->right == NULL) {
            path += to_string(root->val);
            result.push_back(path);
            return;
        }
        if (root) {
            if (root->left != NULL) {
                dfs(root->left, path + to_string(root->val) + "->", result);
            }
            if (root->right != NULL) {
                dfs(root->right, path + to_string(root->val) + "->", result);
            }
        }
    }
};
